### ISSUE TITLE
Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR)

### ISSUE THREAD

**Issue by salgiza** [realm/realm-dotnet#2456]
## Goals

After upgrading an app to Realm 10.x it stopped to launch in Android... but only for new deployments. Those users that already had an old version of the app installed didn't notice anything (neither did the developers), which has made everything so much fun...

## Expected Results

The app running and users happy.

## Actual Results

The app failing to launch and new users lost.

## Steps to Reproduce

- Download the QuickJournal example from this repository.
- Upgrade the nuget packages to their latest versions: Realm 10.2.0, Xamarin.Forms 5.0.0.2012, Xamarin.Essentials 1.6.1.
- Remove the EmbedAssembliesIntoApk from the Android Debug config (you can just delete the line from QuickJournal.Android.csproj)
- Try to debug the app (either in an emulator or a device). **Important: no data must exist for the app** (either delete the app first or delete its data).
- It will crash with the message in the title.

I haven't tested it in detail, but I think the crash also happens if the app is compiled for x86 (and deployed in an Intel Android image), regardless of the value of EmbedAssembliesIntoApk.

Note that it also works if AndroidFastDeploymentType is enabled.

Some extra info: this was a project that was created a few years ago, and either the EmbedAssembliesIntoApk did not exist then, or was not preconfigured by default (this was not something that we had changed since then). That's the reason why it was not set, and the fact that it actually worked perfectly on those devices that just upgraded the version totally caught us by surprise.

## Version of Realm and Tooling

- Flavor:
  - [ ] MongoDB Realm
  - [x] Local Database only
- Client SDK Version: Realm .Net 10.2.0
- Client OS & Version: Android (crashed on all versions of Android 10.x and Android 11.x that we tested).

**Comment by DominicFrei** [comment]
@salgiza Thank you for that report. I will have a look into it.

**Comment by DominicFrei** [comment]
I did run a couple tests.

Current status: I cannot reproduce the crash on my side yet.

For reference, my changes:
https://github.com/realm/realm-dotnet/tree/df/crash-in-quick-journel

This includes the following devices, all freshly created:
![devices](https://user-images.githubusercontent.com/1942012/122777930-6a034200-d2a4-11eb-8756-2b98a4ebaa67.jpg)

NuGet:
![nuget](https://user-images.githubusercontent.com/1942012/122777934-6a9bd880-d2a4-11eb-81fe-e214ed68ffde.jpg)

@salgiza Just to double check: Are there any other changes that you did that I maybe did not do? 

And another thing: Could you clarify this part?
```
Remove the EmbedAssembliesIntoApk from the Android Debug config
(you can just delete the line from QuickJournal.Android.csproj)
```

Looking into https://github.com/realm/realm-dotnet/blob/master/examples/QuickJournal/QuickJournal.Android/QuickJournal.Android.csproj I cannot find any `EmbedAssembliesIntoApk` at all.

I'll do some more tests in the meantime.

**Comment by salgiza** [comment]
Mmm… that is very weird, because other than upgrading the nugget packages I don’t think I did anything else. I did those steps, compiled, checked that it worked in the emulator and compared the csproj with the one in my app.

I won’t be back at the office until Wednesday; as soon as I can get another look at it I’ll update the ticket (I’ll try to do the same steps again).

The only difference I can think of: I was using the latest stable version of VS Studio for macOS. Maybe opening the project (or updating the packages) with it automatically updated some values in the csproj file?

I didn’t even know there was a EmbedAssembliesIntoApk parameter before this morning, so I certainly didn’t add it there intentionally 😮

**Comment by DominicFrei** [comment]
> The only difference I can think of: I was using the latest stable version of VS Studio for macOS. Maybe opening the project (or updating the packages) with it automatically updated some values in the csproj file?

I would hope it to work the same but yeah, this is a difference since I am on Windows.
 
> I didn’t even know there was a EmbedAssembliesIntoApk parameter before this morning, so I certainly didn’t add it there intentionally 😮

See https://docs.microsoft.com/en-gb/xamarin/android/deploy-test/building-apps/build-process#Fast_Deployment.

When you activate `Use Fast Deployment (debug mode only)` in the `Android Options` in the solution's properties it will lead to `<EmbedAssembliesIntoApk>false</EmbedAssembliesIntoApk>` being set.
If you leave it off you'll end up with `<EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>`.

The `EmbedAssembliesIntoApk` does have a default parameter but it seems like it gets explicitely set as soon as I play around with the `Fast Deployment` option. I guess that's why I did not see it at first (see the link to our `master` above).

**Comment by salgiza** [comment]
Ok, I've done it again from the beginning, and I had missed a step in my report: when I downloaded the sample project I upgraded the nuget packages... but then I got a build error the first time I tried to run it: 

![CleanShot 2021-06-23 at 13 24 43](https://user-images.githubusercontent.com/6630306/123088739-65ff2d80-d426-11eb-809f-27320c70b219.png)

So I went to the project properties, and I changed the Target Android version:

![CleanShot 2021-06-23 at 13 25 27](https://user-images.githubusercontent.com/6630306/123088847-89c27380-d426-11eb-9861-20ed29eaab39.png)

And this is where, Visual Studio for macOS, updated the Android.csproj file (this time I initialized a git repository to keep track of all changes):

![CleanShot 2021-06-23 at 13 26 40](https://user-images.githubusercontent.com/6630306/123088982-aced2300-d426-11eb-971d-0ad8c677a95b.png)

Once this step is done, the app crashes on debug, if I do any of the following things.

1. I enable x68 and x86_64 as supported ABIs, to compile for the emulator (by default they are unchecked):

![CleanShot 2021-06-23 at 13 28 13](https://user-images.githubusercontent.com/6630306/123089188-e9b91a00-d426-11eb-9ffa-f920096b7ef1.png)

2. I remove the EmbedAssembliesIntoApk tag from the .csproj file.

In any of these cases, this is the result:

![CleanShot 2021-06-23 at 13 31 17](https://user-images.githubusercontent.com/6630306/123089500-4e747480-d427-11eb-84dd-d409fb1ddfd6.png)

The weird thing (thinking about the fact that it worked for you in Windows) is that on Monday I did try it with another member of the team that uses Windows, and he did get the same crash all of the time, just as customers did (weirdly enough, I can run the emulator even though the x86 ABIs are not selected, but his VS for Windows complained and didn't let him debug the app).

And, just in case, here is a zip file of the Example solution, after enabling the x68 and x86_64 ABIs.
https://www.dropbox.com/s/13hhi36vibg1f7a/QuickJournal.zip

**Comment by dgerding** [comment]
Hi, 
I think I have the same (related?) error, and maybe the way it is cropping will help explain the problem: I ONLY get error when I try and run on the Android Emulator (x86, standard VS2019 latest, not-preview). So it does sound like a build/link related thing? 

Here's the output I could get from Android telemetry. I wrapped the call that crashes the app:
`_realm = Realm.GetInstance(_realmConfiguration);`

in try/catch but never got VS debugger to catch anything :P

(Also, on a hunch I tried with and without encryption key. same result)
-DG 

Log from Android telemetry before crash:

06-25 14:42:33.225: A/crashpad(2338): dlopen: dlopen failed: library "libandroidicu.so" not found: needed by /system/lib/libharfbuzz_ng.so in namespace (default)
06-25 14:42:33.226: A/libc(2176): Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x14 in tid 2176 (companyname.appname), pid 2176 (companyname.appname)
06-25 14:42:33.261: I/crash_dump32(2341): obtaining output fd from tombstoned, type: kDebuggerdTombstone
06-25 14:42:33.263: I/tombstoned(291): received crash request for pid 2176
06-25 14:42:33.263: I/crash_dump32(2341): performing dump of process 2176 (target tid = 2176)
06-25 14:42:33.270: A/DEBUG(2341): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
06-25 14:42:33.270: A/DEBUG(2341): Build fingerprint: 'google/sdk_gphone_x86_arm/generic_x86_arm:11/RSR1.201013.001/6903271:userdebug/dev-keys'
06-25 14:42:33.270: A/DEBUG(2341): Revision: '0'
06-25 14:42:33.270: A/DEBUG(2341): ABI: 'x86'
06-25 14:42:33.270: A/DEBUG(2341): Timestamp: 2021-06-25 14:42:33-0500
06-25 14:42:33.270: A/DEBUG(2341): pid: 2176, tid: 2176, name: companyname.appname  >>> com.companyname.appname <<<
06-25 14:42:33.270: A/DEBUG(2341): uid: 10161
06-25 14:42:33.270: A/DEBUG(2341): signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x14
06-25 14:42:33.270: A/DEBUG(2341): Cause: null pointer dereference
06-25 14:42:33.270: A/DEBUG(2341):     eax b5e30a70  ebx b5d2a220  ecx bfd09004  edx 00000014
06-25 14:42:33.270: A/DEBUG(2341):     edi eba150e8  esi b5e30a70
06-25 14:42:33.270: A/DEBUG(2341):     ebp ff896bc8  esp ff896b40  eip b55f8d23
06-25 14:42:33.314: A/DEBUG(2341): backtrace:
06-25 14:42:33.314: A/DEBUG(2341):       #00 pc 000f8d23  /data/app/~~JREtJ2UMWgMsICQdVqQufA==/com.companyname.appname-HDdf8L-b498BoQx_kFoRQw==/lib/x86/librealm-wrappers.so (shared_realm_get_table_key+45) (BuildId: 5c77e0faf21179a06036a77da9b76266bc4b853e)
06-25 14:42:33.314: A/DEBUG(2341):       #01 pc 0000b86c  <anonymous:b7b0f000>
06-25 14:40:23.820: D/logd(0): logdr: UID=10161 GID=10161 PID=2341 n tail=50 logMask=8 pid=2176 start=0ns timeout=0ns
06-25 14:40:23.824: D/logd(0): logdr: UID=10161 GID=10161 PID=2341 n tail=50 logMask=1 pid=2176 start=0ns timeout=0ns
06-25 14:40:23.955: D/logd(0): logdr: UID=10161 GID=10161 PID=2341 n tail=0 logMask=8 pid=2176 start=0ns timeout=0ns
06-25 14:40:23.957: D/logd(0): logdr: UID=10161 GID=10161 PID=2341 n tail=0 logMask=1 pid=2176 start=0ns timeout=0ns
06-25 14:42:33.599: E/tombstoned(291): Tombstone written to: /data/tombstones/tombstone_26
06-25 14:42:33.601: W/ActivityTaskManager(529):   Force finishing activity com.companyname.appname/crc6467ce5f43874b1979.MainActivity

**Comment by dgerding** [comment]
![image](https://user-images.githubusercontent.com/523279/123481398-47856780-d5c9-11eb-8e8e-5d9b07438568.png)

I can report that turning off Use Fast Deployment fixed the problem for me. Can't promise it's the same problem.

**Comment by nirinchev** [comment]
Hi, we've identified an issue that only manifests when running on x86 devices. We have a fix and hope to release an update soon.

**Comment by salgiza** [comment]
> Hi, we've identified an issue that only manifests when running on x86 devices. We have a fix and hope to release an update soon.

Although this is great, because currently our problem is related to running the app in the x86 emulator, please remember that my original report was because, for us, the app crashed in production (deployed in ARM Android devices) if <EmbedAssembliesIntoApk> was not enabled for the release build. And only if the app had not previously been deployed in the device/the Realm db already existed (this one was quite hard to catch).

Not a problem anymore, because we have enabled that option (as it actually should), but I guess that it is something to take into account.

**Comment by nirinchev** [comment]
Do you have a stack trace from a crash on ARM? Would be curious to see if it's in any way similar to the one on the emulator. Also, do you know what the exact model (or at least the CPU architecture) of the devices that crash is - e.g. have you seen it with both ARMv7 and ARMv8?

**Comment by salgiza** [comment]
Unfortunately I don't, but it was the same error as in the emulator: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR)...

Regarding the devices, looking at the Google Play logs, crashes during the day that the broken app was published came from:
Zebra Technologies TC20
Zebra Technologies TC26
Xiaomi Redmi 6A (cactus)
Samsung SM-J610F

And internally we also reproduced the bug on a Moto G5 Plus and a Zebra TC20.

(I must admit I'm not sure how to check whether those devices are ARMv7 or ARMv8).

I would have thought that you would get the same error if deploying the Quick Journal app in an Android device, following the steps I described, but if it works for you, I can try to reproduce it again using that example.

**Comment by nirinchev** [comment]
Those are arm64 devices and shouldn't suffer from the same problem that was causing the x86 simulator crashes. We'll need to investigate further.

**Comment by DominicFrei** [comment]
> I would have thought that you would get the same error if deploying the Quick Journal app in an Android device, following the steps I described, but if it works for you, I can try to reproduce it again using that example.

@salgiza We were indeed eventually able to reproduce an issue which matches the stack trace posted by @dgerding.
But only when using an Android x86 emulator.

The important part here is:
```
06-25 14:42:33.314: A/DEBUG(2341): #0 pc 000f8d23 
/data/app/~~JREtJ2UMWgMsICQdVqQufA==
/com.companyname.appname-HDdf8L-b498BoQx_kFoRQw==
/lib/x86/librealm-wrappers.so (shared_realm_get_table_key+45)
(BuildId: 5c77e0faf21179a06036a77da9b76266bc4b853e)
```
There is an issue when converting table and object key values which we created a fix for in https://github.com/realm/realm-dotnet/pull/2477.

I wonder if there is another issue hiding behind your crash since turning fast deployment on and off seems to make a difference for you. It did not (and should not) make a difference for above mentioned bug and fix though.

Unfortunately
```
Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR)...
```
could mean anything, it does not mean it's the same error. It's some kind of fatal crash that might or might not be the same crash @dgerding experiences. Without any stack traces or logs it is hard to compare but we need to do more tests and see if we can find anything in addition to the current knowledge.

**Comment by salgiza** [comment]
Ok, it seems I was lucky enough to trigger two bugs (yay!) that crashed the app on launch.

As I debugged the crash using the emulator, this means that what I investigated last Monday was actually the x86 bug and not the actual bug that affected us. But, luckily? enough, the same thing that fixed the emulator bug (enabling EmbedAssembliesIntoApk in the project) fixed the app in production.

So, everything that I reported on this ticket, I guess, was related to the x86 bug and not whatever caused the original crash (thus, I don't even know if Realm is the cause, as I can't debug at which point the app is crashing).

For what it's worth, this is the log that I get when I deploy the app in Release on an Android device, but unless there is something in it that makes you think that this is related to Realm, I think that this ticket can be closed as fixed as soon as the x86 bug is solved.


```
[ResourceType] For resource 0x7f0f015f, entry index(35106-28 13:17:37.127 W/ResourceType(14117): For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For re06-28 13:17:37.128 W/ResourceType(14117): For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f021a, entry index(538) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f01df, entry index(479) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f023c, entry index(572) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0221, entry index(545) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0222, entry index(546) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0239, entry index(569) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f018d, entry index(397) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0235, entry index(565) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0247, entry index(583) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0248, entry index(584) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0220, entry index(544) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f021e, entry index(542) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f014c, entry index(332) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0135, entry index(309) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f014c, entry index(332) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f015f, entry index(351) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0148, entry index(328) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0149, entry index(329) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0251, entry index(593) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0250, entry index(592) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f024e, entry index(590) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f00d2, entry index(210) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0224, entry index(548) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0224, entry index(548) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0226, entry index(550) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f022a, entry index(554) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0228, entry index(552) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f01da, entry index(474) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0238, entry index(568) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f020f, entry index(527) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0211, entry index(529) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0236, entry index(566) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0212, entry index(530) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f01e0, entry index(480) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0213, entry index(531) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0217, entry index(535) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f021a, entry index(538) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f01df, entry index(479) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f021e, entry index(542) is beyond06-28 13:17:37.139 W/ResourceType(14117): For resource 0x7f0f023c, entry index(572) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0221, entry index(545) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0222, entry index(546) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0239, entry index(569) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f023a, entry index(570) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f018d, entry index(397) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0235, entry index(565) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0247, entry index(583) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0248, entry index(584) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0220, entry index(544) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f021e, entry index(542) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f014c, entry index(332) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0135, entry index(309) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f014c, entry index(332) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f015f, entry index(351) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0148, entry index(328) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0149, entry index(329) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0251, entry index(593) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f0250, entry index(592) is beyond type entryCount(179)
[ResourceType] For resource 0x7f0f018b, entry index(395) is beyond type entryCount(179)
Android Language:es-ES
.NET Language/Locale:es-ES
[libc] Fatal signal 11 (SIGSEGV), code 1, fault addr 0xfffffffffffff1 in tid 14117 (amics.warehouse)
```

**Comment by nirinchev** [comment]
If you're able to repro that on device, can you attach the logcat output around the crash (there should be messages similar to pc#0 some-address) - if we're lucky, we should be able to see the last native call before the segfault. If it's realm_get_table_key with some numbers and letters in between, it's almost certain to be the same bug as the one discovered on x86. If it's something else with Realm in it, it's likely something not yet known. And if it doesn't include realm at all, it's probably some other library.

**Comment by salgiza** [comment]
This is what I get using logcat (I didn't know VS for macOS didn't show the whole log?!):

```
06-28 14:41:32.116 15744 15744 V mono-stdout: Android Language:es-ES
06-28 14:41:32.119 15744 15744 V mono-stdout: .NET Language/Locale:es-ES
06-28 14:41:32.912 15744 15744 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0xfffffffffffff1 in tid 15744 (amics.myapp-id)
06-28 14:41:32.913   409   409 W         : debuggerd: handling request: pid=15744 uid=10159 gid=10159 tid=15744
06-28 14:41:32.989 15773 15773 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
06-28 14:41:32.989 15773 15773 F DEBUG   : Build fingerprint: 'Zebra/TC20/TC20RT:7.1.2/01-08-21-0-NG-00-M1/170921:user/release-keys'
06-28 14:41:32.989 15773 15773 F DEBUG   : Revision: '0'
06-28 14:41:32.989 15773 15773 F DEBUG   : ABI: 'arm64'
06-28 14:41:32.989 15773 15773 F DEBUG   : pid: 15744, tid: 15744, name: amics.myapp-id  >>> com.company.myapp-id <<<
06-28 14:41:32.989 15773 15773 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xfffffffffffff1
06-28 14:41:32.989 15773 15773 F DEBUG   :     x0   0000000000000014  x1   0000007fee040e68  x2   0000000000000014  x3   0000000020203166
06-28 14:41:32.989 15773 15773 F DEBUG   :     x4   ffffffffffffffff  x5   0000000000000000  x6   0000000080808080  x7   6565656565656565
06-28 14:41:32.989 15773 15773 F DEBUG   :     x8   0000000000000000  x9   9e7ff636abb97dd1  x10  7f7f7f7f7f7f7f7f  x11  0101010101010101
06-28 14:41:32.989 15773 15773 F DEBUG   :     x12  0000000000000020  x13  000000008000002f  x14  0000007fafb1c428  x15  0000007fafb1c0cc
06-28 14:41:32.989 15773 15773 F DEBUG   :     x16  0000007f90e5a520  x17  0000007fafae5b34  x18  0000000000000008  x19  0000000000000000
06-28 14:41:32.989 15773 15773 F DEBUG   :     x20  0000000000000000  x21  0000000000000040  x22  0000007f90e32f58  x23  0000007f90d92f15
06-28 14:41:32.989 15773 15773 F DEBUG   :     x24  0000007f90daf6b7  x25  000000000000002e  x26  0000007f90dad681  x27  0000007f90d92f13
06-28 14:41:32.989 15773 15773 F DEBUG   :     x28  fffffffffffffff1  x29  0000007fee0412f0  x30  0000007f90d610a8
06-28 14:41:32.990 15773 15773 F DEBUG   :     sp   0000007fee041290  pc   0000007f90d610b8  pstate 0000000080000000
06-28 14:41:32.991 15773 15773 F DEBUG   : 
06-28 14:41:32.991 15773 15773 F DEBUG   : backtrace:
06-28 14:41:32.991 15773 15773 F DEBUG   :     #00 pc 00000000002740b8  /data/app/com.company.myapp-id/lib/arm64/libmonosgen-2.0.so (mono_dump_mem+144)
06-28 14:41:32.991 15773 15773 F DEBUG   :     #01 pc 00000000002740a4  /data/app/com.company.myapp-id/lib/arm64/libmonosgen-2.0.so (mono_dump_mem+124)
06-28 14:41:33.449  1291 15776 W ActivityManager:   Force finishing activity com.company.myapp-id/crc6490831dd6f0c5d996.MainActivity
```

Just for privacy I changed the app id, but it does not seem to reference realm.

**Comment by nirinchev** [comment]
Ugh, that looks like a segfault in mono itself 😬 it's unlikely we'll be able to do anything about it, but as long as changing the build settings resolves it, I guess it's alright. And we'll release a x86 fix shortly so you should be able to resume testing on the emulator.

**Comment by ryanholden8** [comment]
Hi @salgiza, I'm trying to follow the conclusions made in this thread. Was your team able to fix these `Fatal Signal 11` - `libmonosgen-2.0.so` crashes for release build/ARM devices by setting `EmbedAssembliesIntoApk` to `false` ?

Our team is also facing these crashes in production. Specifically with a Zebra TC20 as well.

Our latest approach was to change the [mono GC bridge implementation from tarjan to new](https://github.com/xamarin/xamarin-android/issues/2929#issuecomment-480874971) since the crash is always proceeded by a garbage collection cycle. Still waiting to see reports from production on how effective that change was.

**Comment by salgiza** [comment]
> Hi @salgiza, I'm trying to follow the conclusions made in this thread. Was your team able to fix these `Fatal Signal 11` - `libmonosgen-2.0.so` crashes for release build/ARM devices by setting `EmbedAssembliesIntoApk` to `false` ?
> 
> Our team is also facing these crashes in production. Specifically with a Zebra TC20 as well.
> 
> Our latest approach was to change the [mono GC bridge implementation from tarjan to new](https://github.com/xamarin/xamarin-android/issues/2929#issuecomment-480874971) since the crash is always proceeded by a garbage collection cycle. Still waiting to see reports from production on how effective that change was.

At the end of the day, and after doing extra testing when this ticket was closed, we found that the crashes in the version we published in Google Play were due to... switching git branches.

We had to release a small bugfix for the stable version of the app and, in order to do so, I moved from the dev branch into an older branch (the version published in Google Play). I applied the fix, I checked that it worked (in debug mode), compiled for release, and published the aab file.

Now, what happened there? It seems that VS for macOS had cached something (from the dev branch), somewhere, and the release version crashed on launch. Cleaning the project (and manually deleting all bin an obj folders) fixed the problem. And I surely did this during one of my many tests when trying to find the source of the crash (in the simulator), so I wrongly assumed that the only changed that I had done to the project (EmbedAssembliesIntoApk=true in release, false in debug) had fixed the error.

It probably had to do with an updated nuget package in the dev branch (it might had been realm, but I would certainly not consider this a bug on them), and we were able to reproduce it by doing the same again: compile dev, change branch to an older version, compile in release -> crash on launch. Clean/delete folders -> fixed.

I don't know if this will be the same that is happening to you (anyway, clean/delete all bin/obj folders just in case), but I hope that this information helps someone!
